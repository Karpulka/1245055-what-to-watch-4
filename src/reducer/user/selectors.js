import NameSpace from "../name-space";

const NAME_SPACE = NameSpace.USER;

export const getAuthorizationStatus = (state) => {
  return state[NAME_SPACE].authorizationStatus;
};

export const getErrorMessage = (state) => {
  return state[NAME_SPACE].errorMessage;
};

export const getIsEmailValid = (state) => {
  return state[NAME_SPACE].isEmailValid;
};
