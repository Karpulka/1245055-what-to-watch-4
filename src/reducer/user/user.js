import {setNewObject} from "../../utils";

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  errorMessage: ``,
  isEmailValid: true
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_ERROR_TEXT: `SET_ERROR_TEXT`,
  SET_EMAIL_VALID: `SET_EMAIL_VALID`
};

const ErrorText = {
  EMAIL_IS_INVALID: `Please enter a valid email address`,
  TRY_AGAIN: `We canâ€™t recognize this email and password combination. Please try again.`
};

const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status
  }),

  setError: (message) => ({
    type: ActionType.SET_ERROR_TEXT,
    payload: message
  }),

  setEmailValidation: (message) => ({
    type: ActionType.SET_EMAIL_VALID,
    payload: message
  })
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch((err) => {
        throw err;
      });
  },

  login: (authData) => (dispatch, getState, api) => {
    const emailValid = authData.login.match(/^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i);

    if (!authData.login || (emailValid && !authData.password)) {
      dispatch(ActionCreator.setError(ErrorText.TRY_AGAIN));
      return false;
    }

    if (!emailValid) {
      dispatch(ActionCreator.setEmailValidation(ErrorText.EMAIL_IS_INVALID));
      return false;
    }

    return api.post(`/login`, {
      email: authData.login,
      password: authData.password
    })
      .then(() => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
      })
      .catch(() => {
        dispatch(ActionCreator.setError(ErrorText.TRY_AGAIN));
      });
  }
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return setNewObject(state, {
        authorizationStatus: action.payload,
        isEmailValid: true,
        errorMessage: ``
      });

    case ActionType.SET_ERROR_TEXT:
      return setNewObject(state, {
        errorMessage: action.payload,
        isEmailValid: true
      });

    case ActionType.SET_EMAIL_VALID:
      return setNewObject(state, {
        errorMessage: action.payload,
        isEmailValid: false
      });
  }

  return state;
};

export {ActionCreator, ActionType, AuthorizationStatus, Operation, reducer};
