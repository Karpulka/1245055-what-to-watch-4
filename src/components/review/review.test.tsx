import * as React from "react";
import * as renderer from "react-test-renderer";
import Review from "./review";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import NameSpace from "../../reducer/name-space";
import {AuthorizationStatus} from "../../reducer/user/user";
import {StaticRouter} from "react-router-dom";

const mockStore = configureStore([]);

const film = {
  id: 2,
  title: `Test film`,
  background: `/some/bg.jpg`,
  src: `/some/poster.jpg`,
};

it(`Render Review Component is Auth`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      isDisableCommentForm: false,
      errorText: ``
    },
    [NameSpace.USER]: {
      authorizationStatus: AuthorizationStatus.AUTH
    }
  });

  const {title, background, src, id} = film;

  const tree = renderer
    .create(<Provider store={store}>
      <StaticRouter>
        <Review
          title={title}
          background={background}
          src={src}
          id={id}
          isDisableSubmit={true}
          onSubmitReview={() => {}}
          onChangeText={() => {}}
          isDisableForm={false}/>
      </StaticRouter>
    </Provider>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});

it(`Render Review Component is No Auth`, () => {
  const store = mockStore({
    [NameSpace.DATA]: {
      isDisableCommentForm: false,
      errorText: ``
    },
    [NameSpace.USER]: {
      authorizationStatus: AuthorizationStatus.NO_AUTH
    }
  });

  const tree = renderer
    .create(<Provider store={store}>
      <StaticRouter>
        <Review
          title={film.title}
          background={film.background}
          src={film.src}
          id={film.id}
          isDisableSubmit={true}
          onChangeText={() => {}}
          onSubmitReview={() => {}}
          isDisableForm={false}/>
      </StaticRouter>
    </Provider>)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
