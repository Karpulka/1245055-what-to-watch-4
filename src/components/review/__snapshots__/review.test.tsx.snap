// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Review Component is Auth 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Test film"
        src="/some/bg.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/2"
              onClick={[Function]}
            >
              Test film
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/"
              onClick={[Function]}
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      
      <div
        className="user-block"
      >
        <div
          className="user-block__avatar"
        >
          <a
            href="/mylist"
            onClick={[Function]}
          >
            <img
              alt="User avatar"
              height="63"
              src="/img/avatar.jpg"
              width="63"
            />
          </a>
        </div>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="The Grand Budapest Hotel poster"
        height="327"
        src="/some/poster.jpg"
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    
    <form
      action="#"
      className="add-review__form"
    >
      <fieldset
        disabled={false}
      >
        <div
          className="rating"
        >
          <div
            className="rating__stars"
          >
            <input
              className="rating__input"
              defaultChecked={true}
              defaultValue={1}
              id="star-0"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-0"
            >
              Rating 1
            </label>
            <input
              className="rating__input"
              defaultChecked={false}
              defaultValue={2}
              id="star-1"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-1"
            >
              Rating 2
            </label>
            <input
              className="rating__input"
              defaultChecked={false}
              defaultValue={3}
              id="star-2"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-2"
            >
              Rating 3
            </label>
            <input
              className="rating__input"
              defaultChecked={false}
              defaultValue={4}
              id="star-3"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-3"
            >
              Rating 4
            </label>
            <input
              className="rating__input"
              defaultChecked={false}
              defaultValue={5}
              id="star-4"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-4"
            >
              Rating 5
            </label>
          </div>
        </div>
        <div
          className="add-review__text"
        >
          <textarea
            className="add-review__textarea"
            id="review-text"
            name="review-text"
            onChange={[Function]}
            placeholder="Review text"
          />
          <div
            className="add-review__submit"
          >
            <button
              className="add-review__btn"
              disabled={true}
              onClick={[Function]}
              type="submit"
            >
              Post
            </button>
          </div>
        </div>
      </fieldset>
    </form>
  </div>
</section>
`;

exports[`Render Review Component is No Auth 1`] = `
<section
  className="movie-card movie-card--full"
>
  <div
    className="movie-card__header"
  >
    <div
      className="movie-card__bg"
    >
      <img
        alt="Test film"
        src="/some/bg.jpg"
      />
    </div>
    <h1
      className="visually-hidden"
    >
      WTW
    </h1>
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <nav
        className="breadcrumbs"
      >
        <ul
          className="breadcrumbs__list"
        >
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/films/2"
              onClick={[Function]}
            >
              Test film
            </a>
          </li>
          <li
            className="breadcrumbs__item"
          >
            <a
              className="breadcrumbs__link"
              href="/"
              onClick={[Function]}
            >
              Add review
            </a>
          </li>
        </ul>
      </nav>
      
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__poster movie-card__poster--small"
    >
      <img
        alt="The Grand Budapest Hotel poster"
        height="327"
        src="/some/poster.jpg"
        width="218"
      />
    </div>
  </div>
  <div
    className="add-review"
  >
    
    <form
      action="#"
      className="add-review__form"
    >
      <fieldset
        disabled={false}
      >
        <div
          className="rating"
        >
          <div
            className="rating__stars"
          >
            <input
              className="rating__input"
              defaultChecked={true}
              defaultValue={1}
              id="star-0"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-0"
            >
              Rating 1
            </label>
            <input
              className="rating__input"
              defaultChecked={false}
              defaultValue={2}
              id="star-1"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-1"
            >
              Rating 2
            </label>
            <input
              className="rating__input"
              defaultChecked={false}
              defaultValue={3}
              id="star-2"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-2"
            >
              Rating 3
            </label>
            <input
              className="rating__input"
              defaultChecked={false}
              defaultValue={4}
              id="star-3"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-3"
            >
              Rating 4
            </label>
            <input
              className="rating__input"
              defaultChecked={false}
              defaultValue={5}
              id="star-4"
              name="rating"
              onChange={[Function]}
              type="radio"
            />
            <label
              className="rating__label"
              htmlFor="star-4"
            >
              Rating 5
            </label>
          </div>
        </div>
        <div
          className="add-review__text"
        >
          <textarea
            className="add-review__textarea"
            id="review-text"
            name="review-text"
            onChange={[Function]}
            placeholder="Review text"
          />
          <div
            className="add-review__submit"
          >
            <button
              className="add-review__btn"
              disabled={true}
              onClick={[Function]}
              type="submit"
            >
              Post
            </button>
          </div>
        </div>
      </fieldset>
    </form>
  </div>
</section>
`;
